### ISSUE TITLE
roadmap(offpolicy): 多卡数据并行（N learner × N collector × N GPU，逐 optimizer step 梯度平均）

### ISSUE THREAD

**Issue by TATP-233** [Motphys/UniLab#964]
> Work type: roadmap / infra / performance
> Area: offpolicy / IPC / multi-GPU
> Scope: Standard（5 个 child issues，集成分支 `feat/issue-964-offpolicy-multi-gpu-dp`，main 全程不动）

## Owner summary

将 off-policy 单卡拓扑（#956 终态：1 learner + 1 collector + 1 GPU）扩展为**单节点多卡数据并行**：每张 GPU 一个 learner，每个 learner 绑定一个 collector，replay buffer 按 learner 隔离，learner 之间在每个实际 optimizer step 前通过 flat-gradient all-reduce（torch.distributed / NCCL）做梯度平均，产出 1 个策略。最终 support `sac`（FastSAC）与 `flashsac` + MuJoCo。

> **架构终态修订（2026-08-16）**：原 roadmap 的“周期性参数平均 + `training.dp_sync_interval`”提案已被集成分支的最终实现显式取代。终态语义为 startup state broadcast + 每个实际 optimizer step 前的梯度平均；配置面不再包含 `training.dp_sync_interval`。这不是对原参数平均方案的等价命名。

入口示例：`uv run train --algo sac --task g1_walk_flat --sim mujoco training.devices=[0,1]`。

**与 #956 的关系（显式修订，非静默恢复）**：本 roadmap 修订 #956 冻结决策第 6 条（"不引入多 learner/NCCL、跨 learner replay、参数同步或多 GPU support claim"），并推翻 #937 对多 GPU 能力的移除结论。`tests/algos/test_offpolicy_double_buffer_runner.py:103-119` 的 tombstone 测试随 child 1 同步修订。replay 隔离、learner-owned inference、同卡严格分时、collector 无 actor/CUDA 等其余冻结决策**全部保留**。

## 架构决策

1. **拓扑**：`training.devices=[d0..dN-1]`（新配置字段，默认 `null` = 现状单卡）。主进程为 rank 0 learner，spawn N-1 个 learner 进程；每个 learner 进程通过现有 `AsyncRunner` lifecycle 持有恰好 1 个 collector。不绕开 runner lifecycle，不引入 torchrun 依赖。
2. **训练语义（最终修订）**：rank 间通过 startup state broadcast 保持初始模型与 optimizer state 一致；之后在每个实际 optimizer step 前，把 actor / critic 的稳定 flat-gradient buffer 做 all-reduce mean，再由各 rank 执行 optimizer step。各 rank env / exploration seed 独立，各自从独占 replay 采样；不提供 `training.dp_sync_interval`。
3. **replay 隔离**：每个 learner 的 GPU replay 完整驻留自有显卡，容量、ingress、显存预算校验逻辑不变；不做跨 learner 数据交换。
4. **环境数量**：per-collector `num_envs` 不变（沿用 owner YAML），总 env 数 = N × num_envs。
5. **CPU 分配**：冷路径按 `cpu_count // N` 为每个 collector 自动均分核区，经 #959 已交付的 `EnvCfg.cpu_ids` 透传到 BatchEnvPool；learner 线程预算配套调整。支持显式配置覆盖。
6. **checkpoint / 日志**：rank 0 独占 ExperimentTracker 与 checkpoint；其余 rank 仅上报 metrics。产出策略以 rank 0 checkpoint 为准。
7. **默认行为不变**：不配 `training.devices` 时与现状逐位等价（N=1 退化路径有测试锁定）。

数据流：

```text
GPU_i: learner_i (actor_i, replay_i, 严格分时 inference)
          ↑↓ SharedInferenceSlot / bounded ingress（现有机制，不变）
CPU 分区_i: collector_i (num_envs, 无 actor 无 CUDA)

learner_0..N-1 ←── 每个实际 optimizer step: all-reduce 平均梯度 (NCCL) ──→ 产出单一策略
```

## Child issues（已创建，均挂为 sub-issue）

1. **#965 拓扑与配置面**：`training.devices` 配置 + rank spawn/rendezvous + per-rank (learner+collector) 对 + rank0 日志/checkpoint 归属 + tombstone 测试修订 + N=1 退化等价测试。本 child 不含参数同步（ranks 独立跑通生命周期即可）。
2. **#966 CPU 自动均分接入**：冷路径分区计算 → `EnvCfg.cpu_ids` 透传 + 线程预算配套 + 绑定验证测试。
3. **#967 梯度同步**：NCCL process group（file/tcp rendezvous）、startup state broadcast、每个实际 optimizer step 前的 actor / critic flat-gradient all-reduce mean、seed 策略与同步正确性测试。
4. **#968 多卡 benchmark 与验收**：scaling benchmark（延续 `benchmark/` 惯例）、吞吐与训练曲线验收、支持矩阵与文档更新。
5. **#969 扩展到 flashsac**：前四个 child 完成后，复用全部机制打通 flashsac 多卡路径 + smoke / 回归。

依赖序：#965 → (#966, #967) → #968 → #969。执行方式：每个 child 在独立分支实现，PR 目标为集成分支 `feat/issue-964-offpolicy-multi-gpu-dp`，验收通过后合并，全程不改 main。

## 验收标准

- 正确性：每次同步后各 rank 参数一致；N=1 与现状单卡路径行为等价；collector 依旧无 actor/CUDA（manifest 证明）。
- 性能：开发平台为 2× RTX 6000D（85GB），实测验收以 N=2 为准：稳态 env steps/s ≥ 单卡 1.7×，训练曲线（return vs wall-clock）不退化；4 卡 ≥3× 保留为更大型平台的参考目标，本机无法实测。
- 隔离：各 collector BatchEnvPool 绑定到互不重叠的 cpu_ids 分区（测试证明）。
- 生命周期：任一侧失败沿现有 runner lifecycle 退出，进程组无泄漏。

## Non-goals

- 不做跨 learner 共享 replay / Ape-X 式集中 replay；不做多机多卡。
- 不接入 PyTorch DDP wrapper；梯度同步由现有 runner 在每个实际 optimizer step 通过稳定 flat-gradient all-reduce 完成。
- 首轮不做 td3 / motrix 的 support claim；flashsac 由 child 5（#969）在前四个 child 完成后扩展。
- 不改变 NpEnv / backend / env contract；collector 依旧无 actor、不初始化 CUDA。
- 不自动探测或默认启用多卡；不配 `training.devices` 即现状单卡。
- 拓扑 / 分区规则不放进 `scripts/`，留在 runner / ipc / config 层。

## 规模与维护成本

- 净新增：1 个配置字段（`training.devices`）、1 个 rank launcher（ipc 层）、1 个梯度同步模块、1 个 CPU 分区模块；复用 #956/#959 全部既有机制。
- 长期维护成本：torch.distributed 进程组生命周期与清理路径；多卡正确性/性能测试无法在单卡 CI 跑，需硬件 gate 测试 + N=1 退化路径锁定；支持矩阵新增一行。
- 风险声明：#960 已证明消费级 32 线程机器上 collector 侧无扩展收益；多卡验收依赖目标平台实测。若实测显示 CPU 争用或同步开销导致 scaling 不达标，按实测数据修订范围（同 #956 修订先例）。

## 最终验收结论（2026-08-16）

- **正确性**：采用每个实际 optimizer step 梯度平均后，30 iterations 的 rank0/rank1 checkpoint 77 个同步张量逐位一致（#967 验收）；N=1 退化路径测试锁定；collector 无 actor/CUDA 不变。
- **算法范围**：#969 / #974 已把相同拓扑、同步和生命周期 contract 扩展到 FlashSAC；FastSAC 与 FlashSAC 均完成双卡 smoke。
- **CUDA Graph**：#978 / #979 已恢复 FastSAC 与 FlashSAC 的双卡 optimizer graph capture；eager collective warmup、稳定 flat-gradient buffer 和 graph teardown 均有 reproducer、测试及真机证据。#978 已按 completed 关闭。
- **性能**（2× RTX 6000D + Threadripper 9980X，生产配置 300 iterations）：N=1 稳态 64,242 env steps/s → N=2 聚合 110,804 env steps/s = **1.72×**（≥1.7× 门槛）；final reward 7.96 vs 7.93 不退化；`dp_sync_time` 均值 0.156s/次。数据：`benchmark/outputs/offpolicy_dp_scaling/results.json`。
- **隔离**：/proc 证据 rank0 collector worker 绑定 CPU 0-63、rank1 绑定 64-127，互不重叠。
- **生命周期**：rank 失败经 watchdog 沿 runner lifecycle 退出；正常退出给 sibling ranks 宽限期；无进程泄漏。
- 平台注意：本机 NCCL 的 P2P/SHM 传输不可靠（挂死 / illegal memory access），默认走 TCP loopback（`NCCL_P2P_DISABLE`/`NCCL_SHM_DISABLE` 默认置 1，env 可覆盖）。

- **额外集成项（非原五个 off-policy child）**：#980 基于 Discussion #976 独立完成 RSL-RL PPO 单节点多卡训练，并验证 Go2 joystick flat、G1 motion tracking 的语义、自然退出和吞吐率；#981 将其与最终 shutdown cleanup 合入本集成分支。该项在 umbrella PR 中单独列示，不改写 #964 的 off-policy 验收边界。
- **最终回归**：集成分支最终树通过 `make test-all`：1657 passed、24 skipped、1 xfailed，coverage 72%，benchmark import smoke 通过。

PR 记录：#970（拓扑）、#971（CPU 均分）、#972（同步）、#973（benchmark）、#974（FlashSAC）、#979（NCCL CUDA Graph 与日志）、#980（RSL-RL PPO 多卡）、#981（最终集成与退出清理）、#982（umbrella integration → `main`）。
